fix: Add support for sending large files (up to 2GB)

opssemnik opened the pull request:
# PR Details

Use an additional evaluate function to setup a browser variable,
Break the file down into 50MB chunks and append to the browser with a loop of puppeteer evaluates

During sendMessage event, inside the browser, override the attachment's data with the variable previously created

work is based out of:
https://github.com/pedroslopez/whatsapp-web.js/compare/main...matricce:whatsapp-web.js:fix/sendBigFiles

Updated patch is from @shirser121:
https://github.com/pedroslopez/whatsapp-web.js/pull/2368/files/c452fe86bb0c5ff3c16863ae77e9e76ac8b66010

## Description

Changes sendMessage in Client.js 
it does not affect messages without media, or messages whose media is below 50 Megabytes

Create client objects (chunks) of 50MB parts of the original file contents.
Create metadata chunk to save the amount of data chunks

Afterwards, initiate sendMessage on the client, using a dummy MessageMedia object, which is overwritten inside the browser with the chunks sent previously

## Related Issues

Fixes #2117

## Motivation and Context

Currently the lib only allows <80MB files, which was fine, however, whatsapp web added support for 2GB files (2097152000 bytes)

## How Has This Been Tested
npm install git+https://github.com/opssemnik/whatsapp-web.js#patch-1
Sending / Replying messages without attachments has been tested
Sending / replying messages with media (sticker, non sticker, voice and documents) < 50MB
Sending / replying messages with media (sticker, non sticker, voice and documents) > 50MB < 1GB

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [X] Bug fix (non-breaking change which fixes an issue)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] My code follows the code style of this project.
- [X] I have updated the documentation accordingly (index.d.ts).

opssemnik committed:
Add support for sending large files (up to 1GB)

Break the file down into 50MB chunks and send via batches, creating a string object for each chunk

Afterwards decode each chunk into one bing string

Todo:
I am overriding the entire attachment into a new MessageMedia object. Overriding just the data field was not yielding results (it stayed the original big string)

work is based out of:
https://github.com/pedroslopez/whatsapp-web.js/compare/main...matricce:whatsapp-web.js:fix/sendBigFiles

opssemnik committed:
Update Client.js with latest version from main repo

opssemnik committed:
Fix Large Media messages not opening on iOS

Correctly assign mimetype and filesize values

Santosl2 commented:
In JS we don't use var anymore. Instead of var, try use const for unmodified variables or let for variables that will have modifications.

opssemnik committed:
Update code-style to match standards

opssemnik commented:
> In JS we don't use var anymore (only for very specific cases). Instead of var, try use const for unmodified variables or let for variables that will have modifications.

I've commited the changes, thanks!

opssemnik committed:
Merge branch 'main' into patch-1

shirser121 commented on src/Client.js:
```diff
@@ -815,21 +818,71 @@ class Client extends EventEmitter {
+					window.Store[`mediaChunk_${chatId}_passes`] = passes;	
+			}, chatId,passes);
+        }
+
```
I tried to improve is a little:
```js
        const isBigFile = internalOptions.attachment?.data?.length > (1024 * 1024 * 79);

        if (isBigFile) {
			let startDivision = 2
            let middle = internalOptions.attachment.data.length / startDivision;
            let currentIndex = 0;

            
			while (middle > (1024 * 1024 * 50)){
				startDivision += 1
				middle = Math.floor(internalOptions.attachment.data.length / startDivision);
			}
            
            const randomId = Util.generateHash(32);
            
			while(currentIndex < internalOptions.attachment.data.length){
				let chunkPiece = middle
				if(currentIndex + middle > internalOptions.attachment.data.length){
					chunkPiece = internalOptions.attachment.data.length - currentIndex
				}
				await this.pupPage.evaluate(async (chatId, chunk, randomId) => {
					if (chunk && window[`mediaChunk_${randomId}`]) {
						window[`mediaChunk_${randomId}`] += chunk;
					}
                    else {
                        window[`mediaChunk_${randomId}`] = chunk;
                    }
				}, chatId, internalOptions.attachment.data.substring(currentIndex, currentIndex+chunkPiece), randomId);
				currentIndex += chunkPiece

			}
            
            internalOptions.attachment = new MessageMedia(internalOptions.attachment.mimetype,`mediaChunk_${randomId}`, internalOptions.attachment.filename,internalOptions.attachment.filesize);
        }
```

opssemnik replied:
Fun history
I tried this initially and still crashed..., and forgot about it (i dont have much knowledge on js and puppeteer)
Later i figured that my problem was actually the attachment data not getting overwritten if setting it directly (so i changed to instantiante the whole object into a new MessageMedia)

Later once everything worked, i completely forgot that my previous logic was correct and i had to change it back

Good catch, i will push your changes and modify the PR

Thanks!

shirser121 commented on src/Client.js:
```diff
@@ -815,21 +818,71 @@ class Client extends EventEmitter {
+					currPass += 1;			
+				}
+				delete window.Store[`mediaChunk_${chatId}_passes`];
+            }
```
```js
            if(options?.attachment?.data?.startsWith('mediaChunk')) {
				options.attachment.data = window[options.attachment.data];
				delete window[options.attachment.data];
            }
```

shirser121 reviewed (commented):
I tried to improve the code a little

opssemnik committed:
Update Logic by @shirser121

Change from multiple variables of 50MBs each into one variable that gets appended by 50MB chunks

See:
https://github.com/pedroslopez/whatsapp-web.js/pull/2368/files/c452fe86bb0c5ff3c16863ae77e9e76ac8b66010

shirser121 commented:
Please fix ESLint

opssemnik committed:
Fix ESLint

opssemnik commented:
> Please fix ESLint

Fix commited!

tuyuribr commented on src/Client.js:
```diff
@@ -763,6 +763,7 @@ class Client extends EventEmitter {
             console.warn('Mentions with an array of Contact are now deprecated. See more at https://github.com/pedroslopez/whatsapp-web.js/pull/2166.');
             options.mentions = options.mentions.map(a => a.id._serialized);
         }
+        
```
revert blank spaces

opssemnik replied:
Fixed

tuyuribr commented on src/Client.js:
```diff
@@ -815,21 +818,64 @@ class Client extends EventEmitter {
+
+        const isBigFile = internalOptions.attachment?.data?.length > (1024 * 1024 * 79);
+
+        if (isBigFile) {
```
why not `internalOptions.attachment?.data?.length > (1024 * 1024 * 79)`  ?

opssemnik replied:
Updated

opssemnik committed:
Remove single use variable

shirser121 commented:
Please fix ESLint

opssemnik committed:
Fix ESlint after latest commits

opssemnik commented:
> Please fix ESLint

Fix commited (2)

shirser121 committed:
Merge branch 'main' into patch-1

shirser121 committed:
Merge branch 'main' into patch-1

shirser121 committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

shirser121 committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

tuyuribr reviewed (approved):
LGTM

opssemnik committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

amiruldev20 commented:
does it work?

Sansekai commented:
@opssemnik After the latest wwebjs update, I can't send files over 50MB again

opssemnik commented:
> does it work?

I have been using since i created it, i use it as an youtube video downloader
Feel free to test by installing this branch
npm install git+https://github.com/opssemnik/whatsapp-web.js#patch-1

opssemnik commented:
> @opssemnik After the latest wwebjs update, I can't send files over 50MB again

Could you share your WWEb version? the code is mostly puppeteer based than lib/wweb, so it shouldnt break unless major changes in sendMessage

for me its working in my latest branch with wweb of 2.2332.15

Can you share the version (client.getWWebVersion()), error log and file size?

Sansekai commented:
@opssemnik now im using WWeb version 2.2333.11

opssemnik commented:
> @opssemnik now im using WWeb version 2.2333.11

Thanks,  i will test later with this version

pc-robelbois commented:
How do you guys do to target specific WWeb versions ?

opssemnik commented:
> @opssemnik now im using WWeb version 2.2333.11

I tested now, and it works but it is much much much more slower sending the message, the weird thing is that its the actual WWeb sendMessage (window.sendMessage) that is hanging, the code itself still works (it still successfully calls window.sendMessage)

I will troubleshoot more to identify what are the changes waweb has done to make it work correctly on newer versions

opssemnik commented:
> How do you guys do to target specific WWeb versions ?

Inside your client initiatilization add these properties:
webVersionCache: {
    type: 'remote',
    remotePath: 'https://raw.githubusercontent.com/wppconnect-team/wa-version/main/html/<version>.html',
}

opssemnik commented:
Converted to draft
Found the issue, but i need to dig deeper,
window.Store.MediaUpload.uploadMedia requires some additional fields on larger files, that the lib is not passing, i was able to make it work hardcoded, but i need to debug on calling the function to add them to the media object

opssemnik committed:
Fix large file ABORT_ERROR

opssemnik committed:
Add fileSize to fromFilePath

its required on >100MB documents, and it does not interefere with <100MB documents

opssemnik committed:
Add mediaKeyInfo node to uploadMedia function

Add missing nodes required in new waweb version

opssemnik commented:
@Sansekai fix pushed to my branch, do note it requires the latest whatsapp web version:
2.2333.11
Older versions are not uploading documents larger than 100mb, i assume its because they dont send the additional nodes that the latest sends (which i added to my patch)

Even them in vanilla versions (outside of waweb.js) cannot send files larger than 100mb, so i did not add a version check to the patch (since its not an issue with the lib)

In any case, i fixed support for the latest version, it is now back to life :)

<img width="302" alt="image" src="https://github.com/pedroslopez/whatsapp-web.js/assets/1689221/eac901cb-28c3-4732-8888-d555e75805db">

opssemnik committed:
Add fileSize to MessageMedia.fromFilePath

opssemnik committed:
Fix cached files crashing

If a file has been recently cached, we should not send mediaKeyInfo

however, if a file is not in the cache, we do need to send mediaKeyInfo

opssemnik commented:
> I will look into into as soon as i can. I found something that can help. `mR.findModule('getUploadLimit')[0].getUploadLimit` can give the max size for media type, and we should use it to check if its possible to upload the media as video for example, and return a error if not...

Thanks!, i will look into it, this probably would make the patch "safer", when it stopped working last time, it was due to the promise being rejected, however that reject was not caught by puppeteer

await window.Store.MediaUpload.uploadMedia got stuck forever... , and further media messages were never sent

I will look into that module to get the current upload limit, it seems this limit is based on the current version of wweb you are running, with a check in place, we could make the return clearer to the developers rather than making code stuck

opssemnik committed:
Add upload limit checking

shirser121 commented on src/util/Injected.js:
```diff
@@ -406,6 +408,10 @@ exports.LoadUtils = () => {
 
+        if(mediaInfo.filesize && maxFileSize > mediaInfo.file){
+            throw new Error('Media size exceeds client upload limit');
+        }
```
Please return information about the size limit for the media type.
And add a function that give an option to check the size limit for media type

opssemnik replied:
Yes!, i am currently testing locally and will push soon

opssemnik replied:
Ok i've added a new function, added to index.d.ts , to client.js (getUploadLimits) and updated the error

opssemnik committed:
Add getUploadLimits to Injected

opssemnik committed:
Add getUploadLimits to Client

opssemnik committed:
Update index.d.ts to add getUploadLimits()

opssemnik committed:
Merge pull request #2 from opssemnik/main

Update patch#1

opssemnik committed:
Fix typo

opssemnik committed:
Update error on media size too large

opssemnik commented:
@shirser121 whats your feeling about this function?
my fear right now is people using mistakenly

for instance, if i do MessageMedia.fromFilePath on a video file, it will auto detect mimeType = video/mp4

If i run getUploadLimits(video/mp4) it will fail, because it looks for messageType, not mimeType

additionally, if i force the message as a document, the right limit would be getUploadLimits("document"), not getUploadLimits("video")

Due to this, i made the function take a string (not a message object), however, i don't know if its the clearest way

opssemnik committed:
Update javadocs on getUploadLimits

shirser121 commented:
Please fix ESLint

shirser121 commented:
Great job!

opssemnik committed:
Fix ESLint in Client.js

opssemnik committed:
Fix ESLint on injected

opssemnik commented:
> Please fix ESLint

Done!

shirser121 commented on src/util/Injected.js:
```diff
@@ -118,6 +119,11 @@ exports.LoadUtils = () => {
 
     };
 
+    window.WWebJS.getUploadLimits = async (mediaType) => {
```
We should check if this function check the size of the file after the compressing that whatsapp do or not.
Because if it's after the compressing, the size of the file can be bigger than the limit

opssemnik replied:
it returns the value before compression (based on the values i took in my browser)

shirser121 commented on src/util/Injected.js:
```diff
@@ -403,23 +413,45 @@ exports.LoadUtils = () => {
             mediaObject,
-            mediaType
+            mediaType,
+            isViewOnce: isViewOnce,
```
You shouldn't use 
```js
isViewOnce: isViewOnce
```
You can use
```js
isViewOnce
```
Same for `uploadOrigin` and `forwardedFromWeb`

shirser121 commented on src/util/Injected.js:
```diff
@@ -403,23 +413,45 @@ exports.LoadUtils = () => {
+                mimetype: mediaData.mimetype,
+                mediaObject,
+                mediaType,
+                isViewOnce: isViewOnce,
```
Here also not use same name twice

opssemnik replied:
Thanks, fixed!

opssemnik commented on src/Client.js:
```diff
@@ -866,21 +868,62 @@ class Client extends EventEmitter {
         }
 
+
+        if (internalOptions.attachment?.data?.length > (1024 * 1024 * 79)) {
```
attachment.size does not get fullfilled if the developer creates his own messagemedia out of a base64 string

shirser121 replied:
> attachment.size does not get fullfilled if the developer creates his own messagemedia out of a base64 string (and it used to not get fullfilled when using MessageMedia.fromFilePath as well, but i updated that function)

So why don't we added a condition that if filesize not exist, set filesize to the base64 length as well?
If we have size parameter, why not use it?

opssemnik replied:
I still get a bit skeptical and try to avoid "changing more than necessary"

I've implemented your idea now, tested on a few medias that i generate from base64, as well as doing the same tests as previously (big files, small files, stickers, videos, audios, documents)

Thanks!

opssemnik committed:
Fix code style

opssemnik committed:
Calculate filesize from data if not provided

opssemnik committed:
Update attachment size check

opssemnik committed:
Fix ESLint

opssemnik committed:
Update parameter name for clearer indication

opssemnik committed:
Fix media on ios (1/2)

opssemnik committed:
Fix media on ios (2/2)

opssemnik committed:
Update fileSize check

opssemnik committed:
Dont skip encryption on mandatory files

opssemnik committed:
Fix undef

opssemnik committed:
Fix ESLint

opssemnik committed:
Fix latest ESLint

opssemnik committed:
Fix waveform on downloaded audio files

opssemnik committed:
Merge pull request #3 from opssemnik/fix-waveform-downloaded-audios

Fix waveform on downloaded audio files

opssemnik committed:
Fix typo + eslint

opssemnik committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

aminobot22 commented:
npm install git+https://github.com/opssemnik/whatsapp-web.js#patch-1

i used this one but its not working for me 

not showing any error

opssemnik committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

PurpShell commented:
> @opssemnik
> 
> > tagging @alechkos as he is the creator of the waveform feature
> 
> it's not me, i just brought it from another lib

I brought waveforms to [Baileys](https://github.com/WhiskeySockets/Baileys/pull/180) too from wppconnect, it is a very clean implementation, what is there to solve? I don't get the issue.

purpshell committed:
Merge branch 'main' into patch-1

shirser121 commented:
`Error: Cannot create a string longer than 0x1fffffe8 characters`
We can't read files bigger than 512 MB, so this is pretty useless.
We should change the MessageMedia options, to allow the creation of a pipeline stream instead of creating a base64 string, so we can read the file in chunks.

@PurpShell @alechkos can you help here, please?

opssemnik commented:
> `Error: Cannot create a string longer than 0x1fffffe8 characters` We can't read files bigger than 512 MB, so this is pretty useless. We should change the MessageMedia options, to allow the creation of a pipeline stream instead of creating a base64 string, so we can read the file in chunks.
> 
> @PurpShell @alechkos can you help here, please?

Currently the library breaks at +-50mb because of puppeteer's evaluate function, we hit node's string limit at 10x times the current limit

Locally, for files larger than that, i changed MessageMedia's data to be an array (and the underlying code to support that), however, since node's string limitation has changed in newer versions, i didnt include it in here

opssemnik committed:
Merge branch 'main' into patch-1

opssemnik committed:
Merge branch 'main' into patch-1

opssemnik committed:
Update Client.js

shirser121 commented:
I tried to send media about 380MB and its not work.
Sometimes crash and sometimes just not sent.
Its seem like there is Missing parameter.
Also, i found out that after 3-4 tries i got out of memory error

RC047 commented:
Maybe we have to look for another way to send large files

pierrecorsini commented:
The fact that we are unsure of: what is exactly the limit of size, the conditions of failure, and the failing behaviour; is for me a real issue of stability.

IMHO this would be great to identify a safe limit of size where the upload is stable; and trigger a (bypassable) exception if the file is bigger.

alechkos committed:
Merge branch 'main' into patch-1

alechkos committed:
Merge branch 'main' into patch-1

shirser121 committed:
Merge branch 'main' into patch-1

alechkos committed:
Merge branch 'main' into patch-1

alechkos committed:
Merge branch 'main' into patch-1

alechkos committed:
Merge branch 'main' into patch-1

tofers commented:
I tried this code. It won't send. 

You need to update the code to the latest version and check it again

shirser121 committed:
Merge branch 'main' into patch-1

shirser121 committed:
Merge branch 'main' into patch-1

pedro-php commented:
It's possible to update this code to the new wwebjs version, the one after the merge with the exodus branch? I'm still getting the same error there...

rodrigoslayertech commented:
Any update? This code doesn't work for the current version of Whatsapp Web 2.3000.10*?

YitziG commented:
Could really use this!

purpshell reviewed (approved):
conflicts please
